File: droid/app/src/main/java/com/storyteller/droid/engine/LlamaEngine.kt
package com.storyteller.droid.engine

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.Closeable
import java.io.File

fun interface NativeTokenCallback {
    fun onText(text: String)
}

interface LlamaNativeRuntime {
    fun load(path: String, contextSize: Int): Long
    fun generate(context: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int): String
    fun generateStreaming(
        context: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int,
        callback: NativeTokenCallback,
    ): Int {
        val text = generate(context, prompt, maxTokens, temperature, seed)
        if (text.isNotEmpty()) callback.onText(text)
        return if (text.isEmpty()) 0 else 1
    }
    fun cancel(context: Long)
    fun unload(context: Long)
    fun info(context: Long): String
}

/** Serial, cancellable owner of exactly one native llama model/context pair. */
class LlamaEngine(
    runtimeOverride: LlamaNativeRuntime? = null,
    private val dispatcher: CoroutineDispatcher = Dispatchers.IO,
) : Closeable {
    companion object {
        const val DEFAULT_CONTEXT_SIZE = 2048
        const val MIN_CONTEXT_SIZE = 512
        const val MAX_CONTEXT_SIZE = 8192
        const val MAX_OUTPUT_TOKENS = 1024
        init {
            // JVM unit tests inject a fake runtime and do not package the .so.
            runCatching { System.loadLibrary("llama_jni") }
        }
    }

    private external fun nativeLoadModel(modelPath: String, contextSize: Int): Long
    private external fun nativeGenerate(contextPtr: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int): String
    private external fun nativeGenerateStreaming(
        contextPtr: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int,
        callback: NativeTokenCallback,
    ): Int
    private external fun nativeCancelGeneration(contextPtr: Long)
    private external fun nativeUnloadModel(contextPtr: Long)
    private external fun nativeGetModelInfo(contextPtr: Long): String

    private val runtime = runtimeOverride ?: object : LlamaNativeRuntime {
        override fun load(path: String, contextSize: Int) = nativeLoadModel(path, contextSize)
        override fun generate(context: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int) = nativeGenerate(context, prompt, maxTokens, temperature, seed)
        override fun generateStreaming(
            context: Long, prompt: String, maxTokens: Int, temperature: Float, seed: Int,
            callback: NativeTokenCallback,
        ) = nativeGenerateStreaming(context, prompt, maxTokens, temperature, seed, callback)
        override fun cancel(context: Long) = nativeCancelGeneration(context)
        override fun unload(context: Long) = nativeUnloadModel(context)
        override fun info(context: Long) = nativeGetModelInfo(context)
    }
    private val nativeMutex = Mutex()
    @Volatile private var contextPtr: Long = 0
    @Volatile var isLoaded: Boolean = false
        private set

    suspend fun loadModel(modelPath: String, contextSize: Int = DEFAULT_CONTEXT_SIZE) = withContext(dispatcher) {
        require(contextSize in MIN_CONTEXT_SIZE..MAX_CONTEXT_SIZE) { "Context size must be $MIN_CONTEXT_SIZE..$MAX_CONTEXT_SIZE" }
        require(File(modelPath).isFile) { "Model file not found: $modelPath" }
        nativeMutex.withLock {
            if (isLoaded) return@withLock
            val pointer = runtime.load(modelPath, contextSize)
            check(pointer != 0L) { "Failed to load model: $modelPath" }
            contextPtr = pointer
            isLoaded = true
            runtime.info(pointer) // Force native metadata access while the context is valid.
        }
    }

    suspend fun generate(
        prompt: String,
        maxTokens: Int = 256,
        temperature: Float = 0.8f,
        seed: Int = 0,
    ): String = withContext(dispatcher) {
        require(prompt.isNotBlank()) { "Prompt must not be blank" }
        require(maxTokens in 1..MAX_OUTPUT_TOKENS) { "maxTokens must be 1..$MAX_OUTPUT_TOKENS" }
        require(temperature in 0.0f..2.0f) { "temperature must be 0.0..2.0" }
        nativeMutex.withLock {
            val pointer = contextPtr
            check(isLoaded && pointer != 0L) { "Model not loaded." }
            try {
                runtime.generate(pointer, prompt, maxTokens, temperature, seed)
            } catch (cancelled: CancellationException) {
                runtime.cancel(pointer)
                throw cancelled
            }
        }
    }

    /** Invoke [onText] from the native token loop; no whole-answer rechunking. */
    suspend fun generateStreaming(
        prompt: String,
        maxTokens: Int = 256,
        temperature: Float = 0.8f,
        seed: Int = 0,
        onText: (String) -> Unit,
    ): Int = withContext(dispatcher) {
        require(prompt.isNotBlank()) { "Prompt must not be blank" }
        require(maxTokens in 1..MAX_OUTPUT_TOKENS) { "maxTokens must be 1..$MAX_OUTPUT_TOKENS" }
        require(temperature in 0.0f..2.0f) { "temperature must be 0.0..2.0" }
        nativeMutex.withLock {
            val pointer = contextPtr
            check(isLoaded && pointer != 0L) { "Model not loaded." }
            try {
                runtime.generateStreaming(
                    pointer, prompt, maxTokens, temperature, seed,
                    NativeTokenCallback { text -> if (text.isNotEmpty()) onText(text) },
                )
            } catch (cancelled: CancellationException) {
                runtime.cancel(pointer)
                throw cancelled
            }
        }
    }

    /** Frozen ordered stream around the native semantic callback boundary. */
    fun stream(
        requestId: String,
        prompt: String,
        maxTokens: Int = 256,
        temperature: Float = 0.8f,
        seed: Int = 0,
    ): Flow<ChunkStreamEvent> = callbackFlow {
        val builder = StreamBuilder(requestId)
        send(builder.started())
        val producer = launch(dispatcher) {
            try {
                val count = generateStreaming(prompt, maxTokens, temperature, seed) { text ->
                    runBlocking { send(builder.text(text)) }
                }
                if (count == -2) {
                    send(builder.cancelled())
                } else if (count < 0) {
                    send(builder.failed(StreamErrorCodes.NATIVE_FAILURE))
                } else {
                    send(builder.completed(mapOf("chunks" to count)))
                }
            } catch (_: CancellationException) {
                withContext(NonCancellable) { trySend(builder.cancelled()) }
            } catch (_: Exception) {
                trySend(builder.failed(StreamErrorCodes.NATIVE_FAILURE))
            } finally {
                close()
            }
        }
        awaitClose {
            cancelGeneration()
            producer.cancel()
        }
    }.buffer(BoundedChunkChannel.DEFAULT_CAPACITY)

    /** Interrupts the native token loop without waiting for the serialized call to return. */
    fun cancelGeneration() {
        val pointer = contextPtr
        if (pointer != 0L) runtime.cancel(pointer)
    }

    suspend fun unloadModel() = withContext(dispatcher) {
        cancelGeneration()
        nativeMutex.withLock { unloadLocked() }
    }

    fun onAppBackgrounded() = close()
    fun onMemoryPressure() = close()

    override fun close() {
        cancelGeneration()
        runBlocking(dispatcher) { nativeMutex.withLock { unloadLocked() } }
    }

    private fun unloadLocked() {
        val pointer = contextPtr
        contextPtr = 0
        isLoaded = false
        if (pointer != 0L) runtime.unload(pointer)
    }
}

File: ios/StoryTeller/Engine/LlamaEngine.swift
import Foundation

enum LlamaLifecycleState: Equatable {
    case unloaded
    case loading
    case loaded
    case generating
    case unloading
}

/// Serial, cancellable owner of exactly one native llama model/context pair.
final class LlamaEngine: @unchecked Sendable {
    static let defaultContextSize: Int32 = 2048
    static let minimumContextSize: Int32 = 512
    static let maximumContextSize: Int32 = 8192
    static let maximumOutputTokens: Int32 = 1024

    private let queue = DispatchQueue(label: "com.storyteller.llama", qos: .userInitiated)
    private let lock = NSLock()
    private var contextPtr: UnsafeMutableRawPointer?
    private var internalState: LlamaLifecycleState = .unloaded

    var state: LlamaLifecycleState { lock.withLock { internalState } }
    var isLoaded: Bool { state == .loaded || state == .generating }

    deinit {
        cancelGeneration()
        queue.sync { unloadOnQueue() }
    }

    func loadModel(path: String, contextSize: Int32 = defaultContextSize) async throws {
        guard Self.minimumContextSize...Self.maximumContextSize ~= contextSize else { throw LlamaError.invalidContextSize }
        guard FileManager.default.fileExists(atPath: path) else { throw LlamaError.fileNotFound(path) }
        if isLoaded { return }
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            queue.async { [self] in
                guard transition(from: [.unloaded], to: .loading) else {
                    continuation.resume(throwing: LlamaError.invalidLifecycle)
                    return
                }
                guard let pointer = native_load_model(path, contextSize) else {
                    setState(.unloaded)
                    continuation.resume(throwing: LlamaError.loadFailed(path))
                    return
                }
                lock.withLock { contextPtr = pointer; internalState = .loaded }
                continuation.resume(returning: ())
            }
        }
    }

    func generate(
        prompt: String,
        maxTokens: Int32 = 256,
        temperature: Float = 0.8,
        seed: Int32 = 0
    ) async throws -> String {
        guard !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw LlamaError.emptyPrompt }
        guard 1...Self.maximumOutputTokens ~= maxTokens else { throw LlamaError.invalidTokenLimit }
        guard (0...2).contains(temperature) else { throw LlamaError.invalidTemperature }
        let cancellation = CancellationFlag()
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                queue.async { [self] in
                    guard transition(from: [.loaded], to: .generating), let pointer = pointer() else {
                        continuation.resume(throwing: LlamaError.notLoaded)
                        return
                    }
                    let result = native_generate(pointer, prompt, maxTokens, temperature, seed)
                    setState(.loaded)
                    if cancellation.isCancelled {
                        if let result { free(result) }
                        continuation.resume(throwing: CancellationError())
                    } else if let result {
                        let text = String(cString: result)
                        free(result)
                        continuation.resume(returning: text)
                    } else {
                        continuation.resume(throwing: LlamaError.generationFailed)
                    }
                }
            }
        } onCancel: { [weak self] in
            cancellation.cancel()
            self?.cancelGeneration()
        }
    }

    func generateStreaming(
        prompt: String,
        maxTokens: Int32 = 256,
        temperature: Float = 0.8,
        seed: Int32 = 0,
        onText: @escaping @Sendable (String) -> Void
    ) async throws -> Int {
        guard !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw LlamaError.emptyPrompt
        }
        guard 1...Self.maximumOutputTokens ~= maxTokens else { throw LlamaError.invalidTokenLimit }
        guard (0...2).contains(temperature) else { throw LlamaError.invalidTemperature }
        let cancellation = CancellationFlag()
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                queue.async { [self] in
                    guard transition(from: [.loaded], to: .generating), let pointer = pointer() else {
                        continuation.resume(throwing: LlamaError.notLoaded)
                        return
                    }
                    let box = Unmanaged.passRetained(NativeTextCallbackBox(onText))
                    let result = native_generate_streaming(
                        pointer, prompt, maxTokens, temperature, seed,
                        storytellerNativeTextCallback, box.toOpaque()
                    )
                    box.release()
                    setState(.loaded)
                    if cancellation.isCancelled || result == -2 {
                        continuation.resume(throwing: CancellationError())
                    } else if result < 0 {
                        continuation.resume(throwing: LlamaError.generationFailed)
                    } else {
                        continuation.resume(returning: Int(result))
                    }
                }
            }
        } onCancel: { [weak self] in
            cancellation.cancel()
            self?.cancelGeneration()
        }
    }

    func stream(
        requestId: String,
        prompt: String,
        maxTokens: Int32 = 256,
        temperature: Float = 0.8,
        seed: Int32 = 0
    ) -> BoundedChunkChannel {
        let channel = BoundedChunkChannel()
        let task = Task {
            let builder = LockedStreamBuilder(requestId: requestId)
            channel.send(builder.started())
            do {
                let count = try await generateStreaming(
                    prompt: prompt, maxTokens: maxTokens, temperature: temperature, seed: seed
                ) { text in
                    channel.send(builder.text(text))
                }
                channel.send(builder.completed(["chunks": count]))
            } catch is CancellationError {
                channel.send(builder.cancelled())
            } catch {
                channel.send(builder.failed(StreamErrorCode.nativeFailure))
            }
            channel.finish()
        }
        channel.onCancellation { [weak self] in
            self?.cancelGeneration()
            task.cancel()
        }
        return channel
    }

    func cancelGeneration() {
        if let pointer = pointer() { native_cancel_generation(pointer) }
    }

    func unloadModel() async {
        cancelGeneration()
        await withCheckedContinuation { continuation in
            queue.async { [self] in unloadOnQueue(); continuation.resume() }
        }
    }

    /// Background and memory-pressure callbacks cannot await; cancellation makes
    /// the token loop return, then this serial queue releases all native memory.
    func suspendForBackground() {
        cancelGeneration()
        queue.async { [self] in unloadOnQueue() }
    }

    func releaseForMemoryPressure() { suspendForBackground() }

    private func unloadOnQueue() {
        guard let pointer = pointer() else { setState(.unloaded); return }
        setState(.unloading)
        native_unload_model(pointer)
        lock.withLock { contextPtr = nil; internalState = .unloaded }
    }

    private func pointer() -> UnsafeMutableRawPointer? { lock.withLock { contextPtr } }
    private func setState(_ value: LlamaLifecycleState) { lock.withLock { internalState = value } }
    private func transition(from allowed: Set<LlamaLifecycleState>, to next: LlamaLifecycleState) -> Bool {
        lock.withLock {
            guard allowed.contains(internalState) else { return false }
            internalState = next
            return true
        }
    }
}

private final class NativeTextCallbackBox: @unchecked Sendable {
    let callback: @Sendable (String) -> Void
    init(_ callback: @escaping @Sendable (String) -> Void) { self.callback = callback }
}

private final class LockedStreamBuilder: @unchecked Sendable {
    private let lock = NSLock()
    private var builder: StreamBuilder

    init(requestId: String) { builder = StreamBuilder(requestId: requestId) }
    func started() -> ChunkStreamEvent { lock.withLock { builder.started() } }
    func text(_ value: String) -> ChunkStreamEvent { lock.withLock { builder.text(value) } }
    func completed(_ usage: [String: Int]) -> ChunkStreamEvent {
        lock.withLock { builder.completed(usage) }
    }
    func failed(_ code: String) -> ChunkStreamEvent { lock.withLock { builder.failed(code) } }
    func cancelled() -> ChunkStreamEvent { lock.withLock { builder.cancelled() } }
}

private func storytellerNativeTextCallback(
    _ text: UnsafePointer<CChar>?, _ length: Int32, _ userData: UnsafeMutableRawPointer?
) {
    guard let text, let userData, length > 0 else { return }
    let bytes = UnsafeRawBufferPointer(start: text, count: Int(length))
    let value = String(decoding: bytes, as: UTF8.self)
    if !value.isEmpty {
        Unmanaged<NativeTextCallbackBox>.fromOpaque(userData).takeUnretainedValue().callback(value)
    }
}

private final class CancellationFlag: @unchecked Sendable {
    private let lock = NSLock()
    private var value = false
    var isCancelled: Bool { lock.withLock { value } }
    func cancel() { lock.withLock { value = true } }
}

private extension NSLock {
    func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock(); defer { unlock() }
        return try body()
    }
}

enum LlamaError: LocalizedError {
    case notLoaded, invalidLifecycle, invalidContextSize, invalidTokenLimit, invalidTemperature, emptyPrompt
    case fileNotFound(String), loadFailed(String), generationFailed

    var errorDescription: String? {
        switch self {
        case .notLoaded: return "No model is loaded."
        case .invalidLifecycle: return "The model is changing lifecycle state."
        case .invalidContextSize: return "Context size is outside the supported mobile range."
        case .invalidTokenLimit: return "Output token limit is outside the supported range."
        case .invalidTemperature: return "Temperature must be between 0 and 2."
        case .emptyPrompt: return "Prompt must not be empty."
        case .fileNotFound(let path): return "Model file not found: \(path)"
        case .loadFailed(let path): return "Failed to load model: \(path)"
        case .generationFailed: return "Text generation failed."
        }
    }
}
